internal/libraries/archive/archive.go

```diff
@@ -0,0 +1,118 @@
+// modify or otherwise use the software for commercial activities involving the
+// Arduino software without disclosing the source code of your own applications.
+// To purchase a commercial license, send an email to license@arduino.cc.
+
+// Package archive handles the library release archive.
+package archive
+
+import (
+	"net/url"
+	"os"
+	"path/filepath"
+	"regexp"
+
+	"github.com/arduino/libraries-repository-engine/internal/configuration"
+	"github.com/arduino/libraries-repository-engine/internal/libraries"
+	"github.com/arduino/libraries-repository-engine/internal/libraries/hash"
+	"github.com/arduino/libraries-repository-engine/internal/libraries/metadata"
+	"github.com/arduino/libraries-repository-engine/internal/libraries/zip"
+)
+
+// Archive is the type for library release archive data.
+type Archive struct {
+	SourcePath string
+	RootName   string // Name of the root folder inside the archive.
+	FileName   string
+	Path       string // Full path of the archive.
+	URL        string // URL the archive will have on the download server.
+	Size       int64
+	Checksum   string
+}
+
+// New initializes and returns an Archive object.
+func New(repository *libraries.Repository, libraryMetadata *metadata.LibraryMetadata, config *configuration.Config) (*Archive, error) {
+	repositoryURLData, err := url.Parse(repository.URL)
+	if err != nil {
+		return nil, err
+	}
+	// e.g., https://github.com/arduino-libraries/Servo.git -> github.com
+	repositoryHost := repositoryURLData.Host
+	// e.g., https://github.com/arduino-libraries/Servo.git -> arduino-libraries
+	repositoryParent := filepath.Base(filepath.Dir(repositoryURLData.Path))
+
+	// Unlike the other path components, the filename is based on library name, not repository name URL.
+	fileName := zipFolderName(libraryMetadata) + ".zip"
+
+	return &Archive{
+		SourcePath: repository.FolderPath,
+		RootName:   zipFolderName(libraryMetadata),
+		FileName:   fileName,
+		Path:       filepath.Join(config.LibrariesFolder, repositoryHost, repositoryParent, fileName),
+		URL:        config.BaseDownloadURL + repositoryHost + "/" + repositoryParent + "/" + fileName,
+	}, nil
+}
+
+// Create makes an archive file according to the data of the Archive object and updates the object with the size and
+// checksum for the resulting file.
+func (archive *Archive) Create() error {
+	err := os.MkdirAll(filepath.Dir(archive.Path), os.FileMode(0755))
+	if err != nil {
+		return err
+	}
+
+	if err := zip.Directory(archive.SourcePath, archive.RootName, archive.Path); err != nil {
+		os.Remove(archive.Path)
+		return err
+	}
+
+	size, checksum, err := getSizeAndCalculateChecksum(archive.Path)
+	if err != nil {
+		return err
+	}
+	archive.Size = size
+	archive.Checksum = checksum
+
+	return nil
+}
+
+// zipFolderName returns the name to use for the folder.
+func zipFolderName(library *metadata.LibraryMetadata) string {
+	pattern := regexp.MustCompile("[^a-zA-Z0-9]")
```

Comment:
This could be declared in the `archive` package scope since it doesn't change so we avoid compiling it each time `zipFolderName` is called.

Reply:
Nice. Added in https://github.com/arduino/libraries-repository-engine/pull/72/commits/7434cd70e41e6003fabf3e2d2575cfd7e9377449

Thanks!